### PR TITLE
GEODE-8147: key commands implement Function-Delta

### PR DESCRIPTION
**Pull request by jhutchison**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/CommandFunction.java [hunk 1]
```diff
@@ -84,10 +88,32 @@ protected Object compute(Region localRegion, ByteArrayWrapper key,
...
-        callable = executeDel(key, localRegion, delType);
+      case DEL: {
+        callable = () -> new RedisKeyInRegion(localRegion, regionProvider).del(key);
         break;
+      }
+      case EXISTS: {
+        callable = () -> new RedisKeyInRegion(localRegion, regionProvider).exists(key);
+        break;
+      }
+//      case RENAME: {
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/CommandFunction.java:99]: remove this deadcode

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameExecutor.java [hunk 2]
```diff
@@ -46,12 +38,22 @@ public void executeCommand(Command command, ExecutionHandlerContext context) {
...
 
-    try (@SuppressWarnings("unused")
-    AutoCloseableLock lockForOldKey = context.getLockService().lock(key)) {
-      try (@SuppressWarnings("unused")
-      AutoCloseableLock lockForNewKey = context.getLockService().lock(newKey)) {
+    List<ByteArrayWrapper> keysToLock = Arrays.asList(key,newKey);
+    keysToLock.sort(ByteArrayWrapper::compareTo);
+
+
+//    try (@SuppressWarnings("unused")
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameExecutor.java:49]: remove this deadcode

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameExecutor.java [hunk 3]
```diff
@@ -95,18 +84,18 @@ public void executeCommand(Command command, ExecutionHandlerContext context) {
...
-            Coder.getErrorResponse(context.getByteBufAllocator(), "Thread interrupted."));
-        return;
-      }
-    } catch (InterruptedException e) {
-      Thread.currentThread().interrupt();
-      command.setResponse(
-          Coder.getErrorResponse(context.getByteBufAllocator(), "Thread interrupted."));
-      return;
-    }
+//      } catch (InterruptedException e) {
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameExecutor.java:87]: remove this deadcode

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/SingleResultRedisFunction.java [hunk 4]
```diff
@@ -40,7 +40,7 @@ public void execute(FunctionContext<Object[]> context) {
         regionFunctionContext.getLocalDataSet(regionFunctionContext.getDataSet());
     Object[] args = context.getArguments();
     RedisCommandType command = (RedisCommandType) args[0];
-    Object result = compute(localRegion, key, command, args);
+    Object result = compute(regionFunctionContext.getDataSet(), key, command, args);
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/SingleResultRedisFunction.java:43]: Why was this change made? getLocalDataSet seems like what we want.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java [hunk 5]
```diff
@@ -46,4 +47,13 @@ public boolean del(ByteArrayWrapper key) {
   public boolean exists(ByteArrayWrapper key) {
     return localRegion.containsKey(key);
   }
+
+  public boolean rename(ByteArrayWrapper oldKey, ByteArrayWrapper newKey) {
+
+    Object value = localRegion.get(oldKey);
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java:53]: What if get returns null? I think in that case rename should immediately return false.

Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java:53]: It looks like this impl will not work on String since they do not yet store their data in the region.
It might be best just to wait for the String work to complete

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameExecutor.java [hunk 6]
```diff
@@ -47,47 +35,15 @@ public void executeCommand(Command command, ExecutionHandlerContext context) {
 
     ByteArrayWrapper key = command.getKey();
     ByteArrayWrapper newKey = new ByteArrayWrapper(commandElems.get(2));
+    RedisKeyCommands redisKeyCommands =
+        new RedisKeyCommandsFunctionExecutor(context.getRegionProvider().getDataRegion());
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameExecutor.java:39]: use AbstractExecutor.getRedisKeyCommands(context) instead

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java [hunk 7]
```diff
@@ -102,4 +102,16 @@ protected RedisData getRedisDataOrDefault(ByteArrayWrapper key, RedisData defaul
...
   }
+
+  public boolean rename(ByteArrayWrapper oldKey, ByteArrayWrapper newKey) {
+    RedisData value = getRedisData(oldKey);
+    if (value == null) {
+      return false;
+    }
+
+    region.put(newKey, value);
+    region.remove(oldKey);
```
Comment (dschneider-pivotal) [inline geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java:113]: If newKey is equal to oldKey it looks like this code will do the wrong thing. The redis rename docs say: "<= 3.2.0: Before Redis 3.2.0, an error is returned if source and destination names are the same.". Perhaps the best way to handle this is to check in the renameExecutor for equal keys and return "OK".

Comment (prettyClouds) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java:113]: we actually added some logic into the RedisKeyCommandsFunctionExecutor for if the key does not exist...we can probably add this here

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameFunction.java [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+import org.apache.geode.redis.internal.executor.set.SingleResultCollector;
+import org.apache.geode.redis.internal.executor.set.StripedExecutor;
+
+@SuppressWarnings("unchecked")
+public class RenameFunction implements Function {
+
+  public static final String ID = "REDIS_RENAME_FUNCTION";
+
+  private final transient StripedExecutor stripedExecutor;
+  private final RegionProvider regionProvider;
```
[the remark this answers is not in the document]
Comment (sabbeyPivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RenameFunction.java:44]: It is! Thank you for catching that.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/StripedExecutor.java [hunk 9]
```diff
@@ -18,10 +18,11 @@
 
 import java.util.concurrent.Callable;
 
+
 /**
  * Allows users to "stripe" their execution in such a way that all tasks belonging to one stripe are
  * executed in-order. A stripe is somehow associated with an Object called the "stripeId".
- * It is upto the implementor of this interface to decide how to do this assocations.
+ * It is up to the implementor of this interface to decide how to do this associations.
```
[the remark this answers is not in the document]
Comment (sabbeyPivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/StripedExecutor.java:25]: good catch, I'll update it.

File: geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java (hunk 7 above)
Comment (sabbeyPivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java:113]: Updating this... are we concerned with Native Redis pre 3.2.0?

Comment (sabbeyPivotal) [reply geode-redis/src/main/java/org/apache/geode/redis/internal/executor/RedisKeyInRegion.java:113]: Updating this and adding a test.

File: geode-redis/src/integrationTest/java/org/apache/geode/redis/general/ShutdownIntegrationTest.java [hunk 10]
```diff
@@ -56,6 +57,7 @@ public void shutdownActuallyShutsDownTheServer() {
   }
 
   @Test
+  @Ignore("will be remedied in: https://github.com/apache/geode/pull/5204")
```
Comment (dschneider-pivotal) [inline geode-redis/src/integrationTest/java/org/apache/geode/redis/general/ShutdownIntegrationTest.java:60]: It would probably be better for you just to wait for: https://github.com/apache/geode/pull/5209. Otherwise this ignore will need to be removed in yet another pr.

Comment (sabbeyPivotal) [reply geode-redis/src/integrationTest/java/org/apache/geode/redis/general/ShutdownIntegrationTest.java:60]: gotcha

Comment (sabbeyPivotal) [reply geode-redis/src/integrationTest/java/org/apache/geode/redis/general/ShutdownIntegrationTest.java]: Merged it in!
